### ISSUE TITLE
Compile CFD code using libmesh (source embeded) failed on ALCF-Mira

### ISSUE THREAD

**Issue by felixit** [libMesh/libmesh#1392]
In fact, this issue is my original issue, which is related with #1391 . I am trying to compile my own CFD code on ALCF-Mira (IBM Blue Gene/Q machines), but always get some errors related with libMesh. My code uses some source files from libMesh. My code is written in C/C++. My code has been compiled on many Linux clusters successfully, but maybe this IBM machines have some special things to be noted. Here is the output of the 'make':

`[ 94%] Built target hpMusic_core
[ 95%] Linking CXX executable ../../../bin/hpMusic_debug
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `libMesh::Xdr::open(std::string const&)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:239: undefined reference to `xdrstdio_create'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<std::basic_string<char> >':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:482: undefined reference to `xdr_string'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `void libMesh::Xdr::data_stream<double>(double*, unsigned int, unsigned int)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:837: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `void libMesh::Xdr::data_stream<float>(float*, unsigned int, unsigned int)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:923: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `void libMesh::Xdr::data_stream<__float128>(__float128*, unsigned int, unsigned int)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:1027: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<int>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<unsigned int>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<short int>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<short unsigned int>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<float>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<double>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<long double>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:505: undefined reference to `xdr_u_int'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:510: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<float>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:519: undefined reference to `xdr_u_int'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<double>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:519: undefined reference to `xdr_u_int'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `xdr_translate<long double>':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:519: undefined reference to `xdr_u_int'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x18): undefined reference to `xdr_int'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x20): undefined reference to `xdr_u_int'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x28): undefined reference to `xdr_short'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x30): undefined reference to `xdr_u_short'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x38): undefined reference to `xdr_float'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o):(.toc+0x40): undefined reference to `xdr_double'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `void libMesh::Xdr::data_stream<int>(int*, unsigned int, unsigned int)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:719: undefined reference to `xdr_vector'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:746: undefined reference to `xdr_vector'
/home/fljia/hpMusic/../hpMesh_gui/debug/lib/libhpMesh_core_debug.a(xdr_cxx.C.o): In function `void libMesh::Xdr::data_stream<unsigned int>(unsigned int*, unsigned int, unsigned int)':
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:719: undefined reference to `xdr_vector'
/home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C:746: undefined reference to `xdr_vector'
/home/fljia/hpMusic/contrib/hdf5/lib/libhdf5.a(H5Z.c.o): In function `H5Z_init_interface':
H5Z.c:(.text+0xa04): undefined reference to `SZ_encoder_enabled'
/home/fljia/hpMusic/contrib/hdf5/lib/libhdf5.a(H5Zszip.c.o): In function `H5Z_filter_szip':
H5Zszip.c:(.text+0x144): undefined reference to `SZ_BufftoBuffDecompress'
H5Zszip.c:(.text+0x1fc): undefined reference to `SZ_BufftoBuffCompress'
collect2: error: ld returned 1 exit status
make[2]: *** [bin/hpMusic_debug] Error 1
make[1]: *** [src/app/hpMusic/CMakeFiles/hpMusic.dir/all] Error 2
make: *** [all] Error 2
`

**Comment by jwpeterson** [comment]
Looks like configure correctly detected that your system does not support XDR:
```
configure:33178: checking for xdrstdio_create
configure:33178: /soft/compilers/wrappers/gcc/mpicxx -o conftest  -std=gnu++11   conftest.cpp -ldl  >&5
/tmp/ccfR7gcL.o: In function `main':
conftest.cpp:(.text+0x14): undefined reference to `xdrstdio_create'
collect2: error: ld returned 1 exit status
configure:33178: $? = 1
```
But then your application code (what is /home/fljia/hpMesh_gui/src/core/mesh/xdr_cxx.C) still uses XDR functions. You might be able to "fix" the issue by wrapping your application in `#ifdef LIBMESH_HAVE_XDR`, but of course you will lose whatever functionality gets wrapped in this way.

**Comment by felixit** [comment]
Hi Peterson,

Really thank you for your help. I am not familiar with the XDR. What is it? What kind of machine can or can't support XDR? This is my first time to compile my code on a IBM Blue Gene HPC.

Best regards,
Feilin Jia

**Comment by jwpeterson** [comment]
> I am not familiar with the XDR. What is it? 

It is a portable binary format that has been around [forever](http://man7.org/linux/man-pages/man3/xdr.3.html). We use it for writing files in libmesh's internal format.

> What kind of machine can or can't support XDR?

Most can. This might be the first one I've seen that doesn't...?

**Comment by jwpeterson** [comment]
Note to self: I've added some idaholab/civet tests for compiling with several features disabled, but `--disable-xdr` does not seem to be one of them. Since we now know about an actual system that doesn't support XDR, I should definitely add that...

**Comment by felixit** [comment]
Thank you for your link. Yes, the Linux should support XDR instinctively. I found an external XDR library, but I am not sure whether it will help [http://people.redhat.com/rjones/portablexdr/](url)

**Comment by jwpeterson** [comment]
It's something that is either on the system or isn't, not something that you can really install yourself. In our configure test for XDR, we don't even link against any particular library as I think it is a part of the OS itself... the only option is to properly guard the code with ifdefs.  And it looks like we might have messed something up in xdr_cxx.C, based on the error messages you posted above...

**Comment by felixit** [comment]
Hi Peterson,

Thank you for your patient help. If it is a part of OS, I think the IBM Blue Gene/Q has some special things of the system, unlike the traditional Linux. I think I will try to contact the support of ALCF to do this. They may encounter the similar problems before. Thank you again.

Best regards,
Felix

**Comment by roystgnr** [comment]
On Mon, 14 Aug 2017, John W. Peterson wrote:

> It's something that is either on the system or isn't, not something
> that you can really install yourself. In our configure test for XDR,
> we don't even link against any particular library as I think it is a
> part of the OS itself...

It's something that should be on the system, in libc... apparently as
of the System V 4.0 standard in *1988*.

That said, there's surely no kernel calls going on inside it, so
there's no reason it *can't* be implemented as a third-party library.
---
Roy

**Comment by felixit** [comment]
Hi Roy,

I found this https://people.redhat.com/~rjones/portablexdr/. I can compile it on my local machine. I will also try this when the ALCF is up soon. 

Best regards,
Felix

**Comment by jwpeterson** [comment]
I could not reproduce any of the problems you reported in xdr_cxx.C, with XDR disabled, so it's possible that you have an old version of xdr_cxx.C in your application. I did find some minor issues with tests that need to be restricted when XDR is not available; those were fixed in #1400. So I'm closing this for now, but feel free to reopen it if you can provide some additional information.
